compiler/dyno/test/resolution/testReturnTypes.cpp

```diff
@@ -0,0 +1,345 @@
+#include "chpl/resolution/scope-queries.h"
+#include "chpl/types/all-types.h"
+#include "chpl/uast/Identifier.h"
+#include "chpl/uast/Module.h"
+#include "chpl/uast/Record.h"
+#include "chpl/uast/Variable.h"
+#include "common.h"
+
+#include <cassert>
+
+struct ReturnVariant {
+  bool literal;
+  std::string intent;
+  std::string type;
+  std::string value;
+  bool generic;
+
+  ReturnVariant(std::string intent, std::string type, bool generic=false)
+    : literal(false), intent(std::move(intent)), type(std::move(type)),
+      value(""), generic(generic) {}
+
+  ReturnVariant(std::string value) : literal(true), intent(""), type(""), value(std::move(value)) {}
+};
+
+std::string buildProgram(std::string intent, std::string selectorValue, const std::vector<ReturnVariant>& variants) {
+  std::ostringstream oss;
+  oss << "proc f(";
+  const char* trail = "";
+  if (selectorValue.empty() && variants.size() > 1) {
+    oss << "sel: int";
+    trail = ", ";
+  }
+  for (size_t i = 0; i < variants.size(); i++) {
+    const auto& variant = variants[i];
+    if (variant.literal) continue;
+    oss << trail << variant.intent << " arg" << i;
+    if (!variant.generic) {
+      oss << ":" << variant.type;
+    }
+    trail = ", ";
+  }
+  oss << ") " << intent << "{" << std::endl;
+
+  if (variants.size() == 1) {
+    // don't need selector if n = 1
+    const auto& variant = variants[0];
+    oss << "  return "
+      << (variant.literal ? variant.value : "arg0") << ";";
+  } else {
+    if (!selectorValue.empty()) {
+      oss << "  param sel = " << selectorValue << ";" << std::endl;
+    }
+    for (size_t i = 0; i < variants.size(); i++) {
+      const auto& variant = variants[i];
+      if (i != 0) {
+        oss << " else ";
+      } else {
+        oss << "  ";
+      }
+      oss << "if " << "__primitive(\"==\", sel, " << i << ") {" << std::endl;
+      oss << "    return ";
+      if (variant.literal) {
+        oss << variant.value;
+      } else {
+        oss << "arg" << i;
+      }
+      oss << ";" << std::endl;
+      oss << "  }";
+    }
+  }
+  oss << std::endl << "}" << std::endl;
+
+  if (selectorValue.empty() && variants.size() > 1) {
+    oss << "var sel : int;" << std::endl;
+  }
+  for (size_t i = 0; i < variants.size(); i++) {
+    const auto& variant = variants[i];
+    if (variant.literal) continue;
+    oss << "var arg" << i << ": " << variant.type << ";" << std::endl;
+  }
+  oss << "var x = f(";
+  trail = "";
+  if (selectorValue.empty() && variants.size() > 1) {
+    oss << "sel";
+    trail = ", ";
+  }
+  for (size_t i = 0; i < variants.size(); i++) {
+    const auto& variant = variants[i];
+    if (variant.literal) continue;
+    oss << trail << "arg" << i;
+    trail = ", ";
+  }
+  oss << ");" << std::endl;
+  return oss.str();
+}
+
+template <typename F>
+void testProgram(std::string intent, std::string selectorValue,
+                 const std::vector<ReturnVariant>& variants, F func) {
```

Comment:
Why does this take in a function as an argument instead of returning a QualifiedType? (IMO returning the QualifiedType would be cleaner).

Reply:
Well, the `Context` argument goes out of scope after the type is returned. Since I use a fresh context for each test, I figure it's cleaner to run a function via `F` before the context is freed. I could, of course, create a new context and pass it in each time, but that will just lead to a wee bit more duplication. I can switch it over if you think that's better

Reply:
Ah, OK. I am used to just passing a `Context*` to all these things. In that case I am fine with the function argument approach for this test. (I care more that you make it call commonType directly).